### ISSUE TITLE
Apparent race condition while updating the API Gateway account

### ISSUE THREAD

**Issue by finferflu** [terraform-providers/terraform-provider-aws#11711]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

`0.11.13`

### Affected Resource(s)

* aws_api_gateway_account

### Terraform Configuration Files

```hcl
# Role attached to API Gateway
resource "aws_iam_role" "api_gw_logging" {
  name = "api-gw-log-role"

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "apigateway.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF
}

# Attach AmazonAPIGatewayPushToCloudWatchLogs policy to API Gateway role to allow it to write logs
resource "aws_iam_role_policy_attachment" "attach_cloudwatch_logging" {
  role       = "${aws_iam_role.api_gw_logging.name}"
  policy_arn = "arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs"
}

# Enable logging for the API Gateway
resource "aws_api_gateway_account" "main" {
  cloudwatch_role_arn = "${aws_iam_role.api_gw_logging.arn}"
}
```

### Expected Behavior

The Terraform configuration should have been applied with no errors.

### Actual Behavior

Got error: 

```
1 error(s) occurred:
* aws_api_gateway_account.main: 1 error(s) occurred:
* aws_api_gateway_account.main: Updating API Gateway Account failed: BadRequestException: The role ARN does not have required permissions configured. Please grant trust permission for API Gateway and add the required role policy.
```

### Steps to Reproduce

1. `terraform apply`

### Important Factoids

I have noticed that if I ran this command after Terraform failed, on the next run, Terraform would succeed:

```
# aws apigateway update-account --patch-operations op='replace',path='/cloudwatchRoleArn',value='arn:aws:iam::xxx:role/api-gw-log-role'
{
    "cloudwatchRoleArn": "arn:aws:iam::xxx:role/api-gw-log-role",
    "throttleSettings": {
        "burstLimit": 5000,
        "rateLimit": 10000.0
    },
    "features": [
        "UsagePlans"
    ],
    "apiKeyVersion": "4"
}
```

If I destroyed the Terraform deployment and ran that command right after, I would get the same error as Terraform:

```
# aws apigateway update-account --patch-operations op='replace',path='/cloudwatchRoleArn',value='arn:aws:iam::xxx:role/api-gw-log-role'
An error occurred (BadRequestException) when calling the UpdateAccount operation: The role ARN does not have required permissions configured. Please grant trust permission for API Gateway and add the required role policy.
```

So it looks like that error only came up when `api-gw-log-role` didn't exist, which made me think something was wrong with the role at the time it's being associated with `aws_api_gateway_account`. By looking at the Terraform build output, I could see that `aws_api_gateway_account` was being created before the `AmazonAPIGatewayPushToCloudWatchLogs` was attached:

```
aws_api_gateway_account.main: Creating...
  cloudwatch_role_arn: "" => "arn:aws:iam::788123154534:role/api-gw-log-role"
  throttle_settings.#: "" => "<computed>"
aws_iam_role_policy_attachment.attach_cloudwatch_logging: Creating...
  policy_arn: "" => "arn:aws:iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs"
  role:       "" => "api-gw-log-role"
```

For this reason, I have updated the `aws_api_gateway_account` configuration to explicitly depend on `aws_iam_role_policy_attachment.attach_cloudwatch_logging`:

```hcl
resource "aws_api_gateway_account" "main" {
  depends_on = ["aws_iam_role_policy_attachment.attach_cloudwatch_logging"] 
  cloudwatch_role_arn = "${aws_iam_role.api_gw_logging.arn}"
}
```

With that change, I consistently get the `BadRequestException` error on the first run, but a success on the next run. This is as far as I could get, so I'd appreciate if you could assist. 

I also want to clarify that I use the same configuration on various projects, but I'm only getting this issue consistently on a single project, which is what I normally use to test configuration changes. 

Thanks!

**Comment by bflad** [comment]
Hi @finferflu 👋 What version of the Terraform AWS Provider are you using?

**Comment by finferflu** [comment]
Hi @bflad, it's `2.45.0`.

**Comment by finferflu** [comment]
@bflad, I forgot to mention that I also rolled back to at some point `2.41.0` and I got the same issue.

**Comment by bflad** [comment]
At first I thought this might be related to #11682 because of the timing, but we're noticing the same in our daily acceptance testing starting 4 days ago (which was a day after #11682 started and we'd be seeing empty error messages instead):

```
--- FAIL: TestAccAWSAPIGatewayAccount_basic (7.77s)
    testing.go:640: Step 0 error: errors during apply:
        
        Error: Updating API Gateway Account failed: BadRequestException: The role ARN does not have required permissions configured. Please grant trust permission for API Gateway and add the required role policy.
```

The `aws_api_gateway_account` resource does attempt to retry for IAM eventual consistency, however it looks like the error messaging changed slightly from:

```
The role ARN does not have required permissions set to API Gateway
```

to

```
The role ARN does not have required permissions configured
```

Would you be interested in patching this? Here's the line that can be adjusted (my recommendation would be to remove ` set to API Gateway` so its still compatible with the older AWS API error message:

https://github.com/terraform-providers/terraform-provider-aws/blob/96840caea28e3abf72d0d994550d2ee7d4b7fc45/aws/resource_aws_api_gateway_account.go#L96

**Comment by finferflu** [comment]
@bflad happy to test, however, I'll need some guidance as I've never worked on the source code for this project. Thanks!

**Comment by bflad** [comment]
@finferflu no worries! Luckily this is a change that only requires an update of the resource logic and not any testing or documentation. Updating that line is sufficient.

We do have a [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md) -- in particular the [running acceptance testing section](https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#running-an-acceptance-test) --that may be helpful, if you wish to ensure the patch works with the existing testing. 😄

**Comment by finferflu** [comment]
Thanks for the guidance, @bflad. This might take a little while as I'm currently busy with various projects, but I'll get back submit a PR soon!

**Comment by finferflu** [comment]
@bflad I was able to submit the patch, thanks for your help!

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 2.46.0 of the Terraform AWS Provider, likely later today. 👍

**Comment by finferflu** [comment]
That's great, thanks @bflad!
